### PR TITLE
civicrm_admin_ui: convert Browse mailing

### PR DESCRIPTION
**Pull request by samuelsov**
Overview
----------------------------------------
Convert Mailing browse page to Search Kit + FormBuilder

Before
----------------------------------------
ksnip_20230614-222956

After
----------------------------------------
ksnip_20230614-201543

Comments
----------------------------------------

Work in progress:
- fix "Created or Sent by"
- make all the route works:
   - action will stop working if we use civicrm/mailing/browse
   - do some extra route to replace default civicrm/mailing/browse/scheduled, civicrm/mailing/browse/unscheduled?
- add language column/filter only if multilingual?
- confirm that the colors for scheduled (green) and paused (red) mailing to make them more obvious are desirable?

### CODE REVIEW

File: ext/civicrm_admin_ui/ang/afsearchMailingBrowseScheduled.aff.html [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<div af-fieldset="">
+  <div class="af-container af-layout-inline">
+    <af-field name="name" />
+    <af-field name="created_date" defn="{input_type: 'Select', search_range: true, options: [{id: '{}', label: 'Choose Date Range'}, {id: 'this.day', label: 'Today'}, {id: 'previous.day', label: 'Yesterday'}, {id: 'ending.week', label: 'Last 7 days including today'}, {id: 'ending.month', label: 'Last 30 days including today'}, {id: 'ending_2.month', label: 'Last 60 days including today'}, {id: 'ending.year', label: 'Last 12 months including today'}, {id: 'this.year', label: 'This calendar year'}, {id: 'previous.year', label: 'Previous calendar year'}, {id: 'ending_2.year', label: 'Last 2 years including today'}, {id: 'ending_3.year', label: 'Last 3 years including today'}, {id: 'next.year', label: 'Next calendar year'}, {id: 'previous_2.fiscal_year', label: 'Previous 2 fiscal years'}, {id: 'previous_before.fiscal_year', label: 'Fiscal year prior to previous fiscal year'}], input_attrs: {}, label: 'Created Date'}" />
```
Comment (larssandergreen) [inline ext/civicrm_admin_ui/ang/afsearchMailingBrowseScheduled.aff.html:4]: There must be a way to load the options directly here.

File: ext/civicrm_admin_ui/ang/afsearchMailingBrowseScheduled.aff.html [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+    <af-field name="from_name" defn="{label: 'Created or Send by', input_attrs: {}}" />
```
Comment (larssandergreen) [inline ext/civicrm_admin_ui/ang/afsearchMailingBrowseScheduled.aff.html:5]: Could we do a contact entity ref select here?

File: ext/civicrm_admin_ui/ang/afsearchMailingBrowseScheduled.aff.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
+{
+    "type": "search",
+    "title": "Find Mailings",
+    "icon": "fa-list-alt",
+    "server_route": "civicrm/mailing/browse/scheduled",
+    "permission": "access CiviCRM",
```
Comment (larssandergreen) [inline ext/civicrm_admin_ui/ang/afsearchMailingBrowseScheduled.aff.json:6]: access CiviMail?

File: ext/civicrm_admin_ui/ang/afsearchMailingBrowseScheduled.aff.html (hunk 2 above)
Comment (larssandergreen) [reply ext/civicrm_admin_ui/ang/afsearchMailingBrowseScheduled.aff.html:5]: Also 'Created or Sen**t** by'

File: ext/civicrm_admin_ui/managed/SavedSearch_Mailing_Browse_Scheduled.mgd.php [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+              'key' => 'created_date',
+              'dataType' => 'Timestamp',
+              'label' => E::ts('Created Date'),
+              'sortable' => TRUE,
+            ],
+            [
+              'type' => 'field',
+              'key' => 'scheduled_id.display_name',
+              'dataType' => 'String',
+              'label' => E::ts('Send By'),
```
Comment (larssandergreen) [inline ext/civicrm_admin_ui/managed/SavedSearch_Mailing_Browse_Scheduled.mgd.php:124]: 'Sen**t** by'

**Comment by larssandergreen** [comment]
Looks good!

Like the colours. I think yellow would make sense for paused rather than red (I might assume that red means it has been cancelled or in some kind of error state), but not a big deal.

Agreed on hiding language if not multilingual. Also need to show Campaign if enabled (search and column).

**Comment by samuelsov** [comment]
Done:
- [x] access CiviMail
- [x] Sent by
- [x] Language and campaign column hidden by default
- [x] yellow line for the paused mailing
- [x] added simplified view for Draft
- [x] removed the `Is archived` filter by having 2 different pages (only one SearchKit but 2 formbuilders)

Remaining tasks:
- [ ]  we need to create alternate actions for disabling, resume, canceling that don't quit the current page, only refresh it
- [ ]  I see that sms use the same search, so we need to ensure it is still working
- [ ]  make `ng-if` work for optional language and campaign filters

File: ext/civicrm_admin_ui/managed/SavedSearch_Mailing_Browse_Scheduled.mgd.php [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+    'label' => E::ts('Started'),
+    'sortable' => TRUE,
+  ],
+  [
+    'type' => 'field',
+    'key' => 'Mailing_MailingJob_mailing_id_01.end_date',
+    'dataType' => 'Timestamp',
+    'label' => E::ts('Completed'),
+    'sortable' => TRUE,
+  ]
```
Comment (colemanw) [inline ext/civicrm_admin_ui/managed/SavedSearch_Mailing_Browse_Scheduled.mgd.php:78]: `civilint` is complaining about this.
```suggestion
  ],
```

**Comment by colemanw** [comment]
> @samuelsov removed the `Is archived` filter by having 2 different pages (only one SearchKit but 2 formbuilders)

What was the reason for this change? From a developer POV it seems simpler to have just one form. Is there a difference from the user POV? (Note that if the reason was to have 2 nav menu items, we **can** set filters from the url so it could be possible to create 2 links to the same form with 2 different filter values set).

**Comment by samuelsov** [comment]
@colemanw no reason that I can remember except trying to replace the 2 existing routes in case it is used elsewhere in the code.

**Comment by mlutfy** [comment]
I committed the syntax fixes reported by Coleman and "Sent By"  reported by Lars.

About the "is archived", I think Samuel meant that we have 3 lists: Drafts, Scheduled/Sent, Archived. It can be the same form, with arguments. The "is archived" option is not really relevant if we keep those 3 menu items/lists.

Then there is the question of whether we replace routes or provide new routes. I've seen sometimes it either one or the other. It's easier to add new routes, but then there may be other areas in the UI that link to specific URLs, so admins would be confused to suddenly be on the old UI?

In other words, in this case, we would be replacing 3 legacy routes by a single route, with URL arguments? (or maybe just 2 routes: Sent vs Archived)

**Comment by colemanw** [comment]
Thanks @mlutfy - I think I can work up a little `hook_civicrm_alterMenu()` magic so that all existing links will point to the new single route. (so "yes" to your question - we can replace all 3 with a single route with arguments).

**Comment by colemanw** [comment]
I'm working on this and have broken out two smaller PRs:
- #28019 
- #28019

**Comment by samuelsov** [comment]
@colemanw A round of testing:

1. Pause/Resume/Cancel actions are not working
2. There is no menu action to Archive a mailing
3. The `is_draft` filter from url doesn't seem to work (there seems to be a value but neither `Yes` nor `No`):
ksnip_20231108-102017
4. After adding a mailing, we go to a generic pages with no selected filter. Ideally, we should keep the current behavior, i.e. :
    - always select Archived = `No` by default
    - if we save a draft, select `Draft` filter to `Yes`
    - if we schedule the mailing, select `Draft` filter to `No`
5. This breaks the `Find Mass Sms` screen as it uses the same url with a parameter `sms=1`.

**Comment by colemanw** [comment]
@samuelsov thanks for testing. Yes drafts are re-broken because we have reverted #28019 in favor of #27559 but that is still waiting for a piece from @totten.
So this PR is very close to finished but I'm going to change it back to draft status for now while we wait for those few missing pieces.

**Comment by mlutfy** [comment]
Any thoughts on how to implement this item?

> "we need to create alternate actions for disabling, resume, canceling that don't quit the current page, only refresh it"

**Comment by magnolia61** [comment]
Slightly related: would it be an idea to call this page Mailing Dashboard? For consistency with the other menu entities?

**Comment by shaneonabike** [comment]
Both of the issues above have been merged now I think. Can we do another round of testing and merge it? @samuelsov @colemanw

**Comment by colemanw** [comment]
I've rebased this and pushed one last update to convert the query & form to use the new `Mailing.status` field. Tested it out and everything crashed! Tracked down the problem to some[ faluty metadata syntax](https://github.com/civicrm/civicrm-core/pull/33561), and with that corrected, now everything seems to work great!
@shaneonabike @magnolia61 @mlutfy @samuelsov @larssandergreen I think we did it!

**Comment by mlutfy** [comment]
Excellent :tada:

I did a bit of poking around, and works for me.

I would add merge-on-pass, but we have to wait for #33561 to be merged first.

**Comment by colemanw** [comment]
That only took 2 years! Full speed ahead! 🚀

**Comment by composerjk** [comment]
In 6.8.0, just noticed the Mailings Sent screen showing triplicates for sent mailings, i.e., showing the same mailing on 3 lines. Might be other issues. Disabled Admin UI for now, as I have work to do.

**Comment by colemanw** [comment]
@composerjk that was fixed with https://github.com/civicrm/civicrm-core/pull/34086
